### ISSUE TITLE
"over" unwraps the top level of unflattened records

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#5275]
## tl;dr

A strict reading of the [`over` operator docs](https://zed.brimdata.io/docs/language/operators/over) indicates the following record should probably be output as is, but instead its top level is broken out into a `key` and `value` pairing.

```
$ echo '{a:{b:{c:1}}}' | zq -z 'over this' -
{key:["a"],value:{b:{c:1}}}
```

## Details

Repro is with Zed commit 21b7168.

Walking through the [`over` docs](https://zed.brimdata.io/docs/language/operators/over), here's how the bullets say `over <expr>` handles different data types:

> an array value generates each of its elements

```
$ zq -version
Version: v1.17.0-55-g21b71680

$ echo '[1,2]' | zq -z 'over this' -
1
2
```

Checks out. ✅ 

> a map value generates a sequence of records of the form `{key:<key>,value:<value>}` for each entry in the map

```
$ echo '|{"APPL":145.03}| |{"GOOG":87.07}|' | zq -z 'over this' -
{key:"APPL",value:145.03}
{key:"GOOG",value:87.07}
```

Checks out. ✅ 

> all other values generate a single value equal to itself.

Noted! Then it goes on to:

> Records can be converted to maps with the [`flatten` function](https://zed.brimdata.io/docs/language/functions/flatten) resulting in a map that can be traversed, e.g., if `this` is a record, it can be traversed with `over flatten(this)`.

```
$ echo '{a:{b:{c:1}}}' | zq -z 'over flatten(this)' -
{key:["a","b","c"],value:1}
```

Putting aside the questionable use of the word "map" here, that also checks out. The path through the hierarchical layers of the record are reflected in the array elements `["a","b","c"]`. ✅ 

But what happens if `over` is attempted on a record _without_ using `flatten`? As a user, my assumption was that this would probably be subject to the "generate a single value equal to itself", but it turns out that's not the case. Instead it unwraps the top level of the record hierarchy into a `key` and puts the remaining levels of the record in a `value`:

```
$ echo '{a:{b:{c:1}}}' | zq -z 'over this' -
{key:["a"],value:{b:{c:1}}}
```

My reason for wanting to do this was that I wanted to operate on a hierarchical record in a [lateral expression](https://zed.brimdata.io/docs/language/lateral-subqueries#lateral-expressions) and hence wanted it to pass through `over` undisturbed. I realized I could get what I want by temporarily treating my record as the single element of an array:

```
$ echo '{a:{b:{c:1}}}' | zq -z 'over [this]' -
{a:{b:{c:1}}}
```

But I hesitate at the thought of showing this in docs as if it's the preferred guidance to users. Indeed, when I shared all of this in a group discussion, the reaction was that the unwrapping of the top level was unexpected, so this should be revisited at some point.

**Comment by philrz** [comment]
`over` was renamed to `unnest` in #6021 and changes in functionality were made at the same time that narrow its functionality specifically to arrays, and with that the exposure to this issue is gone. Below is a summary of where things stand now as of current super commit e214c5f.

Going back to the original repro query, now if `unnest` is attempted on that particular record, it's an error:

```
$ super -version
Version: e214c5f08

$ echo '{a:{b:{c:1}}}' | super -c 'unnest this' -
error({message:"unnest: encountered record without two columns",on:{a:{b:{c:1}}}})
```

As the draft `unnest` docs explain, unpacking a record is done by first applying `flatten` as it was in the past with `over`.

```
$ echo '{a:{b:{c:1}}}' | super -c 'unnest flatten(this)' -
{key:["a","b","c"],value:1}
```

The reference to the "record without two columns" in the error message relates to how, when unpacking arrays as `unnest` is intended for, a record may be used to define how to copy the contents of outer values into the output of `unnest` or a lateral subquery as was done in the past with variables, e.g.,

```
$ echo '{foo:"bar", r:[1,2]}' | super -c 'unnest {other_data_from_outer:foo,unpacked_element:r}' -
{other_data_from_outer:"bar",unpacked_element:1}
{other_data_from_outer:"bar",unpacked_element:2}

$ echo '{foo:"bar", r:[1,2]}' | super -c '
unnest {other_data_from_outer:foo,unpacked_element:r} into (
  values {concat:f"{other_data_from_outer}baz}",times_two:unpacked_element*2}
)' -
{concat:"barbaz}",times_two:2}
{concat:"barbaz}",times_two:4}
```

Running through the other examples that were shown above in the **Details** section, simple unpack of arrays with `unnest`:

```
$ echo '[1,2]' | super -c 'unnest this' -
1
2
```

With map values we've temporarily lost a bit of functionality, since operating on them directly is not possible with `unnest` like it was with `over` since they're not arrays.

```
$ echo '|{"APPL":145.03}| |{"GOOG":87.07}|' | super -c 'unnest this' -
error({message:"unnest: encountered non-array value",on:|{"APPL":145.03}|})
error({message:"unnest: encountered non-array value",on:|{"GOOG":87.07}|})
```

And while applying `flatten` is the way we plan to deal with this, it's not yet possible. This enhancement is tracked in #6050. Until that's taken care of, `flatten` is a no-op on maps, so it won't help with `unnest`.

```
$ echo '|{"APPL":145.03}| |{"GOOG":87.07}|' | super -c 'flatten(this)' -
|{"APPL":145.03}|
|{"GOOG":87.07}|

$ echo '|{"APPL":145.03}| |{"GOOG":87.07}|' | super -c 'unnest flatten(this)' -
error({message:"unnest: encountered non-array value",on:|{"APPL":145.03}|})
error({message:"unnest: encountered non-array value",on:|{"GOOG":87.07}|})
```

Since the original problem can no longer be triggered and the remaining corner case is tracked elsewhere, I'm closing this issue.
